pkg/cloud/metadata_ec2.go

```diff
@@ -1,23 +1,45 @@
 package cloud
 
 import (
+	"context"
 	"fmt"
-	"github.com/aws/aws-sdk-go/aws"
-	"github.com/aws/aws-sdk-go/aws/ec2metadata"
-	"github.com/aws/aws-sdk-go/aws/session"
+
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/feature/ec2/imds"
 	"k8s.io/klog/v2"
 )
 
 type EC2MetadataClient func() (EC2Metadata, error)
 
+type imdsClient struct {
+	*imds.Client
+}
+
+// Available returns if the application has access to the EC2 Metadata service.
+// Can be used to determine if application is running within an EC2 Instance and
+// the metadata service is available.
+//
+// This is taken from v1 of aws-sdk-go to keep the interface the same.
+// See https://github.com/aws/aws-sdk-go/blob/main/aws/ec2metadata/api.go
+func (i *imdsClient) Available() bool {
+	input := &imds.GetMetadataInput{Path: "instance-id"}
+	if _, err := i.GetMetadata(context.Background(), input); err != nil {
+		return false
+	}
+	return true
+}
```

Comment:
To avoid duplicating this code block, can we use the exported function here https://github.com/aws/aws-sdk-go/blob/7112c0a0c2d01713a9db2d57f0e5722225baf5b5/aws/ec2metadata/api.go#L208 instead?

Reply:
I'm not sure we should, the PR is trying to move away from using that Go library and use the v2 version.
In the v2 version, this `Available()` that gets used by the aws-fsx-csi-driver no longer exists, but all it does it precisely this check on `GetMetadata`, so I've implemented it here as a smaller blast radius change, as it is part of an interface that imdsClient needs to implement. A better change could be to change the uses of `Available` in this repo to something else but that felt like it could add too many things to this PR.

Reply:
Hmm, I see. That makes sense, in that case, I suggest getting rid of `Available` altogether. As I understand it, this validation is completely unnecessary. For example, if EC2 IMDS is not available, this will return an error: https://github.com/kubernetes-sigs/aws-fsx-csi-driver/blob/79364b4c9952b28ddabe228c3eed6a01b7792338/pkg/cloud/metadata.go#L80

In other words, lets attempt to retrieve the relevant information from IMDS without the redundant check to see if it is available first with `EC2MetadataInstanceInfo` -- if that returns an error, then we fall back to retrieving the information from K8s.

Reply:
Makes sense, I've just pushed that change and updated the tests. Several of the tests were checking for specific errors from the `EC2MetadataInstanceInfo` call, and as now it's not returning that error I've updated those tests to return a k8s client error like one of the already existing tests does.

Reply:
Great, thanks for the quick turnaround 🙂